lib/edda-server/src/compressor.rs

```diff
@@ -0,0 +1,422 @@
+    #[error("layer db error: {0}")]
+    LayerDb(#[from] LayerDbError),
+    #[error("error while subscribing for messages: {0}")]
+    Subscribe(#[source] StreamError),
+}
+
+type Result<T> = std::result::Result<T, CompressorError>;
+
+#[remain::sorted]
+#[derive(Debug, Clone)]
+enum State {
+    Draining(CompressorUpdateRequestVCurrent),
+    DrainingToRebuild(CompressorRebuildRequestVCurrent),
+    Inactive,
+}
+
+pin_project! {
+    /// A stream that wraps an inner stream with the ability to "compress" requests.
+    pub struct Compressor {
+        incoming: consumer::push::Ordered,
+        state: State,
+        ctx_builder: DalContextBuilder,
+        change_set_id: ChangeSetId,
+        workspace_id: WorkspacePk,
+        ctx: DalContext,
+    }
+}
+
+impl Compressor {
+    pub async fn new(
+        requests_stream: &jetstream::stream::Stream,
+        consumer_config: push::OrderedConfig,
+        ctx_builder: DalContextBuilder,
+        change_set_id: ChangeSetId,
+        workspace_id: WorkspacePk,
+    ) -> Result<Self> {
+        let incoming = requests_stream
+            .create_consumer(consumer_config)
+            .await
+            .map_err(CompressorError::ConsumerCreate)?
+            .messages()
+            .await
+            .map_err(CompressorError::Subscribe)?;
+
+        let ctx = ctx_builder
+            .build_for_change_set_as_system(workspace_id, change_set_id, None)
+            .await
+            .map_err(CompressorError::DalContextBuild)?;
+
+        Ok(Self {
+            incoming,
+            state: State::Inactive,
+            ctx_builder,
+            change_set_id,
+            workspace_id,
+            ctx,
+        })
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum CompressorRequest {
+    CompressorRebuild(CompressorRebuildRequestVCurrent),
+    CompressorUpdate(CompressorUpdateRequestVCurrent),
+}
+
+impl futures::Stream for Compressor {
+    type Item = std::result::Result<Message, consumer::push::MessagesError>;
+
+    fn poll_next(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Option<Self::Item>> {
+        let project = self.project();
+
+        // FIXME(nick): we need to handle timing or max size. What if we have been pending for a long time or have made a massive batch?
+        // We should publish the compressed message at certain limits. We need to first see if there are too many or too few compression
+        // sets happening first though.
+
+        loop {
+            match project.incoming.poll_next_unpin(cx) {
+                Poll::Ready(Some(result)) => {
+                    // First, get the message off the stream and validate it.
+                    let message = match result {
+                        Ok(message) => message,
+                        Err(err) => {
+                            error!(si.error.message = ?err, "error getting info for jetstream message");
```

Comment:
I didn't move errors to the handler because there's a chance naxum isn't the right move after compression. I'm leaning more towards not using naxum for the compressed messages for a few reasons but this is one of them.